File: Deprecated/unix/ehr_geometry/curvature2D.cpp
#include <cmath>
#include <map>
#include <new>
using namespace std;

#include "geoquant.h"
#include "curvature2D.h"
#include "euc_angle.h"
#include "triposition.h"
#include "triangulation.h"

typedef map<TriPosition, Curvature2D*, TriPositionCompare> Curvature2DIndex;
static Curvature2DIndex* Index = NULL;

Curvature2D::Curvature2D( Vertex& v ){    
    angles = new vector<GeoQuant*>();

    GeoQuant* angle;
    for(int i = 0; i < v.getLocalFaces()->size(); i++) {
        angle = EuclideanAngle::At(v, Triangulation::faceTable[(*(v.getLocalFaces()))[i]]);
	    angle->addDependent(this);
	    angles->push_back( angle );
    }
}

void Curvature2D::remove() {
     deleteDependents();
     for(int ii = 0; ii < angles->size(); ii++){
       angles->at(ii)->removeDependent(this);
     }    
     Index->erase(pos);
     delete angles;
     delete this;
}

Curvature2D::~Curvature2D(){ }
void Curvature2D::recalculate() {
    double curv = 2*PI;
    GeoQuant* angle;
    for(int ii = 0; ii < angles->size(); ii++){
      angle = angles->at(ii);
      curv -= angle->getValue();
    }
    
    value = curv;
}

Curvature2D* Curvature2D::At( Vertex& v  ){
  TriPosition T( 1, v.getSerialNumber() );
  if( Index == NULL ) Index = new Curvature2DIndex();
  Curvature2DIndex::iterator iter = Index->find( T );

  if( iter == Index->end() ){
    Curvature2D* val = new Curvature2D( v );
    val->pos = T;
    Index->insert( make_pair( T, val ) );
    return val;
  } else {
    return iter->second;
  }
}

void Curvature2D::CleanUp(){
  if( Index == NULL ) return;
  Curvature2DIndex::iterator iter;
  Curvature2DIndex copy = *Index;
  for(iter = copy.begin(); iter != copy.end(); iter++) {
    iter->second->remove();
  }
    
  delete Index;
  Index = NULL;
}

File: Java/src/Geoquant/Curvature2D.java
package Geoquant;

import java.util.HashMap;
import java.util.LinkedList;

import triangulation.Face;
import triangulation.Triangulation;
import triangulation.Vertex;

import Geoquant.Curvature3D.Sum;

public class Curvature2D extends Geoquant {
  // Index map
  private static HashMap<TriPosition, Curvature2D> Index = new HashMap<TriPosition, Curvature2D>();
  private static Sum total = null;
  // Needed geoquants
  private LinkedList<Angle> angles;
  
  public Curvature2D(Vertex v) {
    super(v);
    
    angles = new LinkedList<Angle>();
    Angle a;
    for(Face f : v.getLocalFaces()) {
      a = Angle.At(v, f);
      a.addObserver(this);
      angles.add(a);
    }
  }
  
  protected void recalculate() {
    value = 2 * Math.PI;

    for(Angle a : angles) {
      value -= a.getValue();
    }
  }
 
  public void remove() {
    deleteDependents();
    for(Angle a : angles) {
      a.deleteObserver(this);
    }
    angles.clear();
    Index.remove(pos);
  }
  
  public static Curvature2D At(Vertex v) {
    TriPosition T = new TriPosition(v.getSerialNumber());
    Curvature2D q = Index.get(T);
    if(q == null) {
      q = new Curvature2D(v);
      q.pos = T;
      Index.put(T, q);
    }
    return q;
  }
  
  public static double valueAt(Vertex v) {
    return At(v).getValue();
  }
  
  public static Sum sum() {
    if(total == null) {
      total = new Sum();
    }
    return total;
  }
  
  public static class Sum extends Geoquant {
    LinkedList<Curvature2D> curvs = new LinkedList<Curvature2D>();
    private static Sum total = null;
    
    private Sum() {
      super();
      Curvature2D k;
      for(Vertex v : Triangulation.vertexTable.values()) {
        k = Curvature2D.At(v);
        k.addObserver(this);
        curvs.add(k);
      }
    }
    protected void recalculate() {
      value = 0;
      for(Curvature2D k : curvs) {
        value += k.getValue();
      }
    }

    public void remove() {
      deleteDependents();
      for(Curvature2D k : curvs) {
        k.deleteObserver(this);
      }
      curvs.clear();
    }
    
    public static Sum get() {
      if(total == null) {
        total = new Sum();
      }
      return total;
    }
    
    public static double getSum() {
      return total.getValue();
    }
  }
}
